src/tools/miri/tests/pass/function_pointers.rs

```diff
@@ -72,10 +76,18 @@ fn main() {
     assert_eq!(indirect3(h), 210);
     assert_eq!(indirect_mut3(h), 210);
     assert_eq!(indirect_once3(h), 210);
-    let g = f as fn() -> i32;
-    assert!(return_fn_ptr(g) == g);
-    assert!(return_fn_ptr(g) as unsafe fn() -> i32 == g as fn() -> i32 as unsafe fn() -> i32);
-    assert!(return_fn_ptr(f) != f);
+    // Check that `i` always has the same address. This is not guaranteed
+    // but Miri currently uses a fixed address for monomorphic functions.
+    assert!(return_fn_ptr(i) == i);
+    assert!(return_fn_ptr(i) as unsafe fn() -> i32 == i as fn() -> i32 as unsafe fn() -> i32);
+    // We don't check anything for `f`. Miri gives it many different addresses
+    // but mir-opts can turn them into the same address.
+    let _val = return_fn_ptr(f) != f;
```

Comment:
This is somewhat dubious. Certainly we want to test that we still successfully give them different addresses sometimes, e.g. by adding `black_box`.

But I am also not convinced that what GVN is doing here is sound. What is the argument for that?

Reply:
I guess if this is something specific to fn-item to fn-ptr coercions, then what the opt does is fine -- that coercion is allowed to deduplicate.

But please fix the test by adding `black_box` and still asserting `!=`. Let's not reduce test coverage unnecessarily.